### PR TITLE
Fix irq guard in `preempt` 

### LINKED ISSUE #478: [Preemption] A Destined Panic in `preempt()`?
# Problem?

After `disable_irq`(in line 68), a `DisabledLocalIrqGuard` is held, which contains a `DisablePreemptGuard` that has increased `PREEMPT_COUNT`.

https://github.com/jinzhao-dev/jinux/blob/0e7f279bd6b1ee4fe564c999d953a5dc1e520bf3/framework/jinux-frame/src/task/processor.rs#L66-L81

https://github.com/jinzhao-dev/jinux/blob/0e7f279bd6b1ee4fe564c999d953a5dc1e520bf3/framework/jinux-frame/src/task/processor.rs#L165-L169

The `disable_irq` is still held when entering `switch_to_task(next_task)` in line 80, and the preemptible check (in line 91) may not succeed?

https://github.com/jinzhao-dev/jinux/blob/0e7f279bd6b1ee4fe564c999d953a5dc1e520bf3/framework/jinux-frame/src/task/processor.rs#L90-L98

https://github.com/jinzhao-dev/jinux/blob/0e7f279bd6b1ee4fe564c999d953a5dc1e520bf3/framework/jinux-frame/src/task/processor.rs#L148-L150

# Solution

If the problem indeed exists, two rudimentary solutions may be:

- drop the `disable_irq` before `switch_to_task(next_task)`;
- [move the preemptible check out from `switch_to_task`](https://github.com/fgh1999/jinux/blob/cf902e58def380b993039d85016148ad8577200d/framework/jinux-frame/src/task/processor.rs#L99-L103), and add [the check](https://github.com/fgh1999/jinux/blob/cf902e58def380b993039d85016148ad8577200d/framework/jinux-frame/src/task/processor.rs#L55-L73) manually before the switch.

### PR DESCRIPTION
**Pull request by fgh1999**
This PR can only fix #478 .
Since #624 won't be merged, and flag `in_preemption` won't be added in `PreemptInfo`, here's a seperate solution for the issue.
There must be potential similiar misbehaviors that can trigger a panic from context switch but have no relation with `preempt`. I guess they will be left to *Atomic mode* and the `might_break`.

### CODE REVIEW

File: framework/aster-frame/src/trap/irq.rs [hunk 1]
```diff
@@ -118,41 +118,62 @@ pub fn disable_local() -> DisabledLocalIrqGuard {
...
+    let was_enabled = irq::is_local_enabled();
+    if was_enabled {
+        irq::disable_local();
+    }
+    was_enabled
+}
+
+/// Enable local IRQs if they are disabled.
+/// Return whether the local IRQs were enabled before this function call.
+pub fn enable_irq_if_disabled() -> bool {
```
Comment (lrh2000) [inline framework/aster-frame/src/trap/irq.rs:141]: When do we need this function? It seems that the return value of this function will never be used, at least in this PR. Is this function just added for consistency with `disable_irq_if_enabled`?

File: framework/aster-frame/src/trap/irq.rs [hunk 2]
```diff
@@ -118,41 +118,62 @@ pub fn disable_local() -> DisabledLocalIrqGuard {
...
     }
 }
 
 impl Drop for DisabledLocalIrqGuard {
     fn drop(&mut self) {
-        if self.was_enabled {
-            irq::enable_local();
+        if self.need_reenable {
+            // Irqs will not be re-enabled if irqs were disabled
+            // when the gaurd is created.
```
Comment (lrh2000) [inline framework/aster-frame/src/trap/irq.rs:175]:
```suggestion
            // IRQs will not be re-enabled if IRQs were disabled
            // when the gaurd is created.
```
I suggest keeping the words in comments consistent and in plain English.

File: framework/aster-frame/src/task/processor.rs [hunk 3]
```diff
@@ -63,7 +67,7 @@ pub fn schedule() {
 
 pub fn preempt() {
     // disable interrupts to avoid nested preemption.
-    let disable_irq = disable_local();
+    disable_irq_if_enabled();
```
Comment (lrh2000) [inline framework/aster-frame/src/task/processor.rs:73]: If the function returns early, the IRQs are disabled forever, which is not correct.

For example, if the function exits here:
```rust
    if !scheduler.should_preempt(&curr_task) {
        return;
    }
```

I also wonder what made you drop your original solution of moving the preempt check from the beginning of `switch_to_task()` to the beginning of `preempt()` and `schedule()`? That solution sounds a bit more correct than the one proposed in this PR. Does it have any problems I'm not aware of?

File: kernel/aster-nix/src/thread/task.rs [hunk 4]
```diff
@@ -31,6 +32,7 @@ pub fn create_new_user_task(user_space: Arc<UserSpace>, thread_ref: Weak<Thread>
             user_mode.context().syscall_ret()
         );
         loop {
+            enable_irq_if_disabled();
```
Comment (lrh2000) [inline kernel/aster-nix/src/thread/task.rs:35]: We should never leave the kernel in a random state where the state of the IRQs is not even known. So this change is also a red flag that implies something wrong happens in the kernel.

File: framework/aster-frame/src/task/processor.rs [hunk 5]
```diff
@@ -141,19 +144,19 @@ impl PreemptInfo {
...
     }
 
     fn is_preemptive(&self) -> bool {
-        self.num_locks.load(Relaxed) == 0
+        self.num_locks.load(Acquire) == 0
     }
 
     fn num_locks(&self) -> usize {
-        self.num_locks.load(Relaxed)
+        self.num_locks.load(Acquire)
```
Comment (lrh2000) [inline framework/aster-frame/src/task/processor.rs:159]: Isn't the preempt count a CPU-local variable where the concurrency access will never occur? I'm not sure why we can't get a mutable reference to CPU-local variables so we can exclusively modify them:
```rust
    fn decrease_num_locks(&mut self) {
        self.num_locks -= 1;
    }
```

File: framework/aster-frame/src/task/processor.rs (hunk 3 above)
Comment (lrh2000) [reply framework/aster-frame/src/task/processor.rs:73]: > I also wonder what made you drop your original solution of moving the preempt check from the beginning of switch_to_task() to the beginning of preempt() and schedule()? That solution sounds a bit more correct than the one proposed in this PR. Does it have any problems I'm not aware of?

OK, I think I know why. We should not really have a non-zero preempt count when we call `switch_task', otherwise the preempt count (as a CPU-local variable) will remain non-zero even after the switch has finished, which is obviously not correct.

So now we have two guards, the preempt guard and the preempt&IRQ guard. The question is whether we need the third guard, the IRQ guard, and if so, we can manually make the preempt&IRQ guard from the preempt guard and the IRQ guard.

File: framework/aster-frame/src/task/processor.rs (hunk 5 above)
Comment (lrh2000) [reply framework/aster-frame/src/task/processor.rs:159]: Thinking deeper. It seems that before having exclusive access to CPU-local variables, we should first disable preempt (otherwise the current task can be preempted and scheduled on another CPU by accident). So it looks like a chicken and egg situation, and solving it seems to be beyond the scope of this PR.

However, I still don't think the fixes in this PR can make a difference. The precise memory order used to preempt the count can be `Relaxed`, since the preempt count is always accessed by a single CPU, and never by other CPUs.

File: framework/aster-frame/src/trap/irq.rs [hunk 6]
```diff
@@ -118,41 +118,62 @@ pub fn disable_local() -> DisabledLocalIrqGuard {
...
+    /// Usually depends on whether the local IRQs were enabled at the moment the guard was created.
+    need_reenable: bool,
     preempt_guard: DisablePreemptGuard,
 }
 
 impl !Send for DisabledLocalIrqGuard {}
 
+/// Disable local IRQs if they are enabled.
+/// Return whether the local IRQs were enabled before this function call.
+pub fn disable_irq_if_enabled() -> bool {
```
Comment (tatetian) [inline framework/aster-frame/src/trap/irq.rs:131]: I have two suggestions.

First, the pair of `disable_irq_if_enabled` and `enable_irq_if_disabled` are easy to misuse compared to `disable_local`. Please keep them as private functions unless there is a strong evidence that the users of the Framework need them.

Second, `disable_irq_if_enabled` is not a particular good name. Technically, the `disable_local` function also disables IRQs only when the IRQs are not disabled already. So the function name `disable_irq_if_enabled` fails to do a good job on distinguishing it from `disable_local`. If the function shall be kept as a public function, I suggest calling it `disable_local_noguard`. If the function is made private, I suggest calling it `do_disable_local`, whose name emphasizes that the function should only be used internally, with caution.

File: framework/aster-frame/src/task/processor.rs [hunk 7]
```diff
@@ -63,7 +70,7 @@ pub fn schedule() {
 
 pub fn preempt() {
     // disable interrupts to avoid nested preemption.
-    let disable_irq = disable_local();
```
Comment (tatetian) [inline framework/aster-frame/src/task/processor.rs:66]: The simplest approach to fix the bug is to remove this line of code, isn't it? The original intention of this line is to make the `preempt` function robust even when our kernel threads can be preempted. But kernel preemption, as opposite to user preemption, is harder to implement. I think for the short term, we should focus on the easier goal of supporting user preemption.

File: framework/aster-frame/src/trap/irq.rs (hunk 6 above)
Comment (lrh2000) [reply framework/aster-frame/src/trap/irq.rs:131]: AFIK, this function is not `disable_local_noguard`. Currently `disable_local` actually does two things, disabling preemption _and_ disabling local IRQs. This function wants to disable IRQs without disabling preemption.

In reality, if local IRQs were disabled, there would be no preemption at all. However, I believe the preemption is still disabled because we want to cause further `schedule()` calls to panic, as it is usually not correct to do task switching when preemption or IRQs are disabled.

`preempt()` is special in the way that it disables IRQs, but it _does_ want to do task switching. Since disabling preemption cause panics in `switch_task()`, this PR introduces a function that disables IRQs without disabling preemption.

This is really confusing and it is hard to learn the difference from the function name. So perhaps we need some better API designs here.

File: framework/aster-frame/src/task/processor.rs (hunk 5 above)
Comment (fgh1999) [reply framework/aster-frame/src/task/processor.rs:159]: To have an exclusive access to CPU-local variables, `CpuLocal::borrow_with` may help by the preempt&IRQ guard at the beginnig.

Not closely related to this PR, will there be risks to replace the `AtomicUsize` with `Cell<usize>` along with `borrow_with`?

Comment (lrh2000) [reply framework/aster-frame/src/task/processor.rs:159]: > To have an exclusive access to CPU-local variables, CpuLocal::borrow_with may help by the preempt&IRQ guard at the beginnig.

The implementation of `CpuLocal` looks like only a placeholder. It needs to be implemented when supporting multiple processors.

> Not closely related to this PR, as for num_locks, will there be risks to replace the AtomicUsize with Cell<usize> along with borrow_with?

I'm not sure about the motivation. Is there any benefit?

`Cell` is [!Sync](https://doc.rust-lang.org/std/cell/struct.Cell.html#impl-Sync-for-Cell%3CT%3E), which means references cannot be shared across threads. This limitation leads to more powerful APIs, so we can get a mutable reference to the inner data.

In contrast, `AtomicUsize`  is [Sync](https://doc.rust-lang.org/std/cell/struct.Cell.html#impl-Sync-for-Cell%3CT%3E). Its references can be shared across threads, resulting in limited APIs. We can never get a mutable reference to the inner data.

As for CPU-local variables, they should not be accessed across threads, so using `Cell` seems to be fine. But I still think both `AtomicUsize` and `Cell` are somehow unnecessary, instead, we should have exclusive access thus a mutable reference `&mut usize` fits better.

More unfortunately, since `borrow_with` disables preemption at the beginning, it causes an infinite recursion if `borrow_with` is used while trying to disable preemption. So it's a chicken and egg situation.

File: framework/aster-frame/src/trap/irq.rs (hunk 6 above)
Comment (tatetian) [reply framework/aster-frame/src/trap/irq.rs:131]: > AFAIK, this function is not `disable_local_noguard`. Currently `disable_local` actually does two things, disabling preemption _and_ disabling local IRQs. This function wants to disable IRQs without disabling preemption.

Yes, you are right. But that doesn't change the fact the function's name is confusing and should be kept private.

> preempt() is special in the way that it disables IRQs, but it does want to do task switching. Since disabling preemption cause panics in switch_task(), this PR introduces a function that disables IRQs without disabling preemption.

My fundamental concern with this PR is that to prevent the panic caused by the preemption guard in `switch_to_task`, the simplest fix to me is to remove the call to `disable_local` in `preempt`. The kernel preemption (as opposite to user preemption) is not the top of our TODO list for improving the scheduling subsystem.

File: framework/aster-frame/src/task/processor.rs (hunk 7 above)
Comment (tatetian) [reply framework/aster-frame/src/task/processor.rs:66]: @fgh1999 Any opinion?

File: framework/aster-frame/src/trap/irq.rs (hunk 6 above)
Comment (lrh2000) [reply framework/aster-frame/src/trap/irq.rs:131]: > My fundamental concern with this PR is that to prevent the panic caused by the preemption guard in switch_to_task, the simplest fix to me is to remove the call to disable_local in preempt. The kernel preemption (as opposite to user preemption) is not the top of our TODO list for improving the scheduling subsystem.

Just removing the `disable_local` call and leaving a TODO here is also fine for me. But this PR and the related issue makes me think about the API design around `disable_local()` and `disable_preempt()`. Perhaps making `disable_local()` also disable preemption is not a good idea. In contrast, we can always do the following:
 - Make `disable_local()` _not_ disable preemption.
 - Add a `might_sleep()` function that checks that both IRQs and preemption are _not_ disabled.
   - Otherwise, we'll panic.
 - Remove the `panic!()` check in `switch_to_task()`.
   - Instead, call `might_sleep()` in the beginning of `preempt()` and `schedule()`.

Compared to the alternatives, i.e., adding the third guard that only disables IRQs without disabling preemption, I think the above proposal achieves a clearer API design.

`might_sleep()` can be useful as it can be called in many other places such as `Mutex::lock()`, to ensure performing the context switching is always safe whenever `Mutex::lock()` is called.

Indeed, I've proposed a similar idea in the original issue last year ([this comment](https://github.com/asterinas/asterinas/issues/478#issuecomment-1807197636)). But it's somehow outside the scope of this PR anyway.

**Comment by fgh1999** [comment]
Grateful for all the advices and information from you two.

This PR will only remove the `disable_local()` in `preempt()`.

File: framework/aster-frame/src/task/processor.rs [hunk 8]
```diff
@@ -62,8 +62,8 @@ pub fn schedule() {
 }
 
 pub fn preempt() {
-    // disable interrupts to avoid nested preemption.
-    let disable_irq = disable_local();
+    // TODO: Refactor `preempt` and `schedule`
+    // after the Atomic mode and `might_break` is enabled.
```
Comment (lrh2000) [inline framework/aster-frame/src/task/processor.rs:66]: We should make comments self-contained. It is not clear what is "Atomic mode" and "might_break" just by reading this comment.

My suggestion, but for references only:
```suggestion
    // TODO: Once kernel preemption is supported, we need to find out a proper way to avoid nested
    // preemption in `schedule()` and `preempt()`. We cannot simply disable preemption here. If we
    // do, the preemption counter will break after a task switch.
```

File: framework/aster-frame/src/trap/irq.rs (hunk 6 above)
Comment (tatetian) [reply framework/aster-frame/src/trap/irq.rs]: I am ok with introducing the atomic mode and `might_break` in another PR. But I am not convinced this thing of "Make disable_local() not disable preemption." Let's discuss it further in the other PR.

File: framework/aster-frame/src/task/processor.rs (hunk 8 above)
Comment (tatetian) [reply framework/aster-frame/src/task/processor.rs:66]: @lrh2000  Good suggestion. But let's do the improvement in the atomic mode PR. I am merging this PR.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
